### PR TITLE
Check for empty value in get_order_id_by_order_key()

### LINKED ISSUE #40055: Check for empty value in get_order_id_by_order_key()
See C0E1AV8T0-slack-p1693994820601269

I was looking into slow backend issues for a friend, and noticed the following query stuck in server:

> SELECT post_id FROM wp_postmeta WHERE meta_key = '_order_key' AND meta_value = ''

I noticed a 3PD plugin wasn't checking for empty order key value before calling `wc_get_order_id_by_order_key()` which likely resulted in a slow query and made the site unresponsive. The site's postmeta table is considerably large.

https://github.com/woocommerce/woocommerce/blob/5a8ed71edc380ac755e1d7e65335897023d01e15/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L449

While plugins should always be careful about this, I am just opening this issue to see if we can do a check before firing the SQL.

Or did I miss anything? Is the empty value already checked in core somewhere?

### PR DESCRIPTION
**Pull request by masteradhoc**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
Check for empty value in get_order_id_by_order_key()
This is especially necessary due to the function being able to call directly see https://github.com/woocommerce/woocommerce/issues/40055#issuecomment-1710260022

Reason is that with an empty variable the sql call gets prepared like this:
`SELECT post_id FROM wp_postmeta WHERE meta_key = '_order_key' AND meta_value = ''`

The issue author mentioned:
_I noticed a 3PD plugin wasn't checking for empty order key value before calling wc_get_order_id_by_order_key() which likely resulted in a slow query and made the site unresponsive. The site's postmeta table is considerably large._

Closes #40055 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Note that this is performance change to prevent unnecessary query, so functionality isn't affected.

Codebase with PR

Start with posts authoritative:

Create a new order and note down it's order key, you can find it by clicking on the customer payment page in the order admin screen and then checking the URL -> &key=`wc_order_RHYQ7nzjy0Xd3`. It's also available on the URL bar in the thank you page after order checkout.
Then in shell -> 
1. call function **wc_get_order_id_by_order_key** with parameter $order_key (valid orderkey) -> should get the order ID
2. call function **wc_get_order_id_by_order_key** with _empty_ parameter $order_key (valid orderkey) -> should return 0 & query shouldn't run

Codebase with**out** PR
3. call function ** wc_get_order_id_by_order_key** with _empty_ parameter $order_key (valid orderkey) -> should return empty & query runs with empty orderkey

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [x] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Check if $order_key is empty in get_order_id_by_order_key() before running the sql query.

</details>

### CODE REVIEW

**Review by lsinger — CHANGES_REQUESTED** [review]
Thank you so much for your contribution, @masteradhoc! ☺️

Could we reverse the flow here to exit early in the non-happy path, and also add a warning? 

As in -- if the order key is empty, print a warning using `_doing_it_wrong( __FUNCTION__, __( '$order_key is empty.', 'woocommerce' ), '8.2.0 )`, and then return `0`. 

The normal case / happy path of returning the query results would then come after that. 

Finally, this feels like a unit test could be a good way to assure this behavior. Please let me know if you need any help writing one.

**Comment by barryhughes** [comment]
Please feel free to re-request review once any feedback is addressed and you are happy for us to take another look :-)

File: plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php
Comment (vedanshujain) [inline plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php]: this should be the first line in the functin.

**Review by vedanshujain — COMMENTED** [review]
This looks good, can you please rebase this branch with the latest trunk so that CI passes (additionally please add a changelog). Thanks!

Additionally, we can just return 0 when the kye is empty, no need to call `_diong_it_wrong`.

**Review by vedanshujain — COMMENTED** [review]
Seems like there are couple more things we need to address before it can be merged:
1. Please use tabs instead of spaces for indents as per repo's lint rules.
2. We also need a changelog, seems like the automated task is failing for some reason. but you can add one manually by using the ` pnpm --filter=./plugins/woocommerce changelog add` command or adding one in `woocommerce/plugins/woocommerce/changelog` folder.

**Comment by vedanshujain** [comment]
Hey @masteradhoc , seems like we are still using spaces instead of tabs for indentation, which is failing the lint check. Would you be able to fix this? Otherwise this PR is good to merge.

**Comment by masteradhoc** [comment]
> Hey @masteradhoc , seems like we are still using spaces instead of tabs for indentation, which is failing the lint check. Would you be able to fix this? Otherwise this PR is good to merge.

@vedanshujain done :)

**Comment by vedanshujain** [comment]
(Waiting for CI pass) Note that I also checked how the function is being used in the extensions, the common pattern is to immediately pass the return value to a `if ( ! $order_id )` or a `wc_get_order` function so this should be safe backward compatibility wise.

**Review by vedanshujain — APPROVED** [review]
LGTM!

Screenshot 2024-10-15 at 11 55 48 AM

**Pull request merged**
